sstables/sstables_manager.cc

```diff
@@ -74,16 +75,26 @@ future<> storage_manager::stop() {
 }
 
 future<> storage_manager::update_config(const db::config& cfg) {
-    for (auto [ep, ecfg] : cfg.object_storage_config()) {
-        auto s3_cfg = make_lw_shared<s3::endpoint_config>(std::move(ecfg));
-        auto [it, added] = _s3_endpoints.try_emplace(ep, std::move(s3_cfg));
-        if (!added) {
+    // Updates S3 client configurations if the endpoint is already known and
+    // removes the entries that are not present in the new configuration.
+    // Even though we remove obsolete S3 clients from this map, each IO
+    // holds a shared_ptr to the client, so the clients will be kept alive for
+    // as long as needed. 
+    std::unordered_map<sstring, s3_endpoint> updates;
+    for (auto& e : cfg.object_storage_endpoints()) {
+        auto s3_cfg = make_lw_shared<s3::endpoint_config>(std::move(e.config));
+        auto existing = _s3_endpoints.find(e.endpoint);
+        if (existing != _s3_endpoints.end()) {
+            auto [it, _] = updates.emplace(e.endpoint, existing->second);
             if (it->second.client != nullptr) {
                 co_await it->second.client->update_config(s3_cfg);
             }
             it->second.cfg = std::move(s3_cfg);
+            continue;
         }
+        updates.emplace(e.endpoint, s3_endpoint(std::move(s3_cfg)));
     }
+    _s3_endpoints = std::move(updates);
```

Comment:
Consider the scenario

1. endpoints are configured with some mistake
2. an sstable is created (keyspace storage option is in action), client pointer is obtained from the `_s3_endpoints`
3. administrator notices the mistake in config, updates it, sends HUP
4. this code puts some _new_ collection into `_s3_endpoints`
5. sstable from step 2 remains with broken client for good (as per my [other](https://github.com/scylladb/scylladb/pull/22952/commits/f4158f25bd959d83f0581269150da340961dc8d4#r2020741168) comment, `update_config()` is not called on any clients)

Reply:
The `update_config()` call will happen. I'm a bit confused, is this the only concern or there is more to this comment?